File: Net/Patterns/State/After/HasQuarterState.cs
namespace Patterns.State.After
{
    using System;

    internal class HasQuarterState : IState
    {
        private readonly GumballMachine gumballMachine;

        public HasQuarterState(GumballMachine gumballMachine)
        {
            this.gumballMachine = gumballMachine;
        }

        public void Insert()
        {
            Console.WriteLine(MachineMessages.InsertTwice);
        }

        public void Eject()
        {
            Console.WriteLine(MachineMessages.EjectSuccessfully);
            gumballMachine.State = gumballMachine.NoQuarter;
        }

        public void Turn()
        {
            Console.WriteLine(MachineMessages.TurnSuccessfully);
            gumballMachine.State = gumballMachine.Sold;
            gumballMachine.Dispense();
        }

        public void Dispense()
        {
            Console.WriteLine(MachineMessages.DispenseWhenHasQuarter);
        }
    }
}

File: Java/After/src/main/state/HasQuarterState.java
package state;

public class HasQuarterState implements State {
	GumballMachine gumballMachine;
 
	public HasQuarterState(GumballMachine gumballMachine) {
		this.gumballMachine = gumballMachine;
	}
  
	public void insertQuarter() {
		System.out.println(MachineMessages.InsertTwice);
	}
 
	public void ejectQuarter() {
		System.out.println(MachineMessages.EjectSuccessfully);
		gumballMachine.setState(gumballMachine.getNoQuarterState());
	}
 
	public void turnCrank() {
		System.out.println(MachineMessages.TurnSuccessfully);
		gumballMachine.setState(gumballMachine.getSoldState());
		gumballMachine.dispense();
	}

    public void dispense() {
        System.out.println(MachineMessages.DispenseWhenHasQuarter);
    }
}
